RecursiveExtractor/Extractors/TarExtractor.cs

```diff
@@ -46,13 +46,15 @@ public async IAsyncEnumerable<FileEntry> ExtractAsync(FileEntry fileEntry, Extra
             {
                 List<FileEntry> entries = new();
 
-                foreach (TarArchiveEntry tarEntry in archive.Entries)
+                TarEntryCollectionEnumerator tarEntryCollectionEnumerator = new(archive.Entries, fileEntry.FullPath);
+                while (tarEntryCollectionEnumerator.GetNextEntry() != null)
                 {
+                    var tarEntry = tarEntryCollectionEnumerator.CurrentEntry;
                     if (tarEntry.IsDirectory)
                     {
                         continue;
                     }
-                    var fs = new FileStream(TempPath.GetTempFilePath(), FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite, bufferSize, FileOptions.DeleteOnClose);
+                    var fs = StreamFactory.GenerateAppropriateBackingStream(options, tarEntry.Size);
```

Comment:
Can you point out where the result from GenerateAppropriateBackingStream is disposed other places? In this case the Stream will be passed to the `FileEntry` constructor with `passthroughStream` true which means the `FileEntry` will not create an extra copy of the stream and will use this `Stream` object directly for its `Content` property, and is disposed by the finalizer of FileEntry. We can't dispose the `Stream` at this point because that `FileEntry` either going to be returned to the caller or extracted further by additional calls.

https://github.com/microsoft/RecursiveExtractor/blob/7c42cfe62460188190d4d1f58a5f7e7f6f48c2a5/RecursiveExtractor/Extractors/TarExtractor.cs#L149

Reply:
There are a few places (`ZipExtractor`, `GzipExtractor`, `BZip2Extractor` where the `GenerateAppropriateBackingStream` call is part of a `using` statement such as:

```
                    using var fs = StreamFactory.GenerateAppropriateBackingStream(options, zipEntry.Size);
```

Your explanation makes sense and this is good to merge.

Reply:
Ah, so looking at ZipExtractor, it is not using passthroughStream argument to FileEntry, so the stream remains owned by the extractor. I think the rationale here is that after all the entries from the zip are extracted the zip itself is not actually returned to the user. I think there's probably a good argument to harmonize the treatments here - because there is the "return archive on fail" argument which returns the unextracted failed archive it probably makes sense to convert to the scheme used by tar rather than by zip. I think I will leave that for a future PR however.